internal/armada/server/lease.go

```diff
@@ -334,6 +306,236 @@ func (q *AggregatedQueueServer) StreamingLeaseJobs(stream api.AggregatedQueue_St
+
+	// Collect all allowed priorities.
+	priorities := make([]int32, 0)
+	if len(q.schedulingConfig.Preemption.PriorityClasses) > 0 {
+		for _, p := range q.schedulingConfig.Preemption.PriorityClasses {
+			priorities = append(priorities, p)
+		}
+	} else {
+		priorities = append(priorities, 0)
+	}
+
+	// Nodes to be considered by the scheduler.
+	nodes := make([]*schedulerobjects.Node, len(req.Nodes))
+	for i, nodeInfo := range req.Nodes {
+		nodes[i] = schedulerobjects.NewNodeFromNodeInfo(&nodeInfo, req.ClusterId, priorities)
+	}
+
+	// Map queue names to priority factor for all active queues, i.e.,
+	// all queues for which the jobs queue has not been deleted automatically by Redis.
+	queues, err := q.queueRepository.GetAllQueues()
+	if err != nil {
+		return nil, err
+	}
+	priorityFactorByQueue := make(map[string]float64)
+	apiQueues := make([]*api.Queue, len(queues))
+	for i, queue := range queues {
+		priorityFactorByQueue[queue.Name] = float64(queue.PriorityFactor)
+		apiQueues[i] = &api.Queue{Name: queue.Name}
+	}
+	activeQueues, err := q.jobRepository.FilterActiveQueues(apiQueues)
+	if err != nil {
+		return nil, err
+	}
+	priorityFactorByActiveQueue := make(map[string]float64)
+	for _, queue := range activeQueues {
+		priorityFactorByActiveQueue[queue.Name] = priorityFactorByQueue[queue.Name]
+	}
+
+	sched, err := scheduler.NewLegacyScheduler(
+		q.schedulingConfig,
+		req.ClusterId,
+		nodes,
+		q.jobRepository,
+		priorityFactorByActiveQueue,
+	)
+	if err != nil {
+		return nil, err
+	}
+	sched.SchedulingReportsRepository = q.SchedulingReportsRepository
+	sched.MinimumJobSize = req.MinimumJobSize
+
+	// Give Schedule() a 3 second shorter deadline than ctx,
+	// to give it a chance to finish up before ctx is cancelled.
+	if deadline, ok := ctx.Deadline(); ok {
+		var cancel context.CancelFunc
+		ctx, cancel = context.WithDeadline(ctx, deadline.Add(-3*time.Second))
+		defer cancel()
+	}
+
+	// Run the scheduler.
+	jobs, mostRecentSuccessfulJobSchedulingReportByQueue, err := sched.Schedule(ctx, aggregatedUsageByQueue)
+
+	// Update the usage report in-place to account for any leased jobs and write it back into Redis.
+	// This ensures resources of leased jobs are accounted for without needing to wait for jobs to start running.
+	for queue, report := range mostRecentSuccessfulJobSchedulingReportByQueue {
+		if queueClusterUsageReport, ok := clusterUsageReport.ResourcesByQueue[queue]; ok {
+			queueClusterUsageReport.ResourcesByPriority = report.TotalQueueResourcesByPriority
+		}
+		err := q.usageRepository.UpdateClusterQueueResourceUsage(req.ClusterId, clusterUsageReport)
+		if err != nil {
+			logging.WithStacktrace(log, err).Error("failed to update cluster usage")
+		}
+	}
+
+	// If we successfully leased jobs, those have to be sent to the executor,
+	// since the leases will have been written into Redis.
+	if len(jobs) > 0 {
+		if err != nil {
+			logging.WithStacktrace(log, err).Error("failed to schedule jobs")
```

Comment:
I think we should have a better error message here.  I *think* this means that we could schedule *some* jobs, but we encountered an error before we scheduled as many jobs as we wanted.